### PR TITLE
[FEATURE] Creating, Submitting, and Handling IPUMS Data Extracts

### PR DESCRIPTION
**Pull request by TheCedarPrince**
This PR introduces the following features through a variety of modifications and new functionalities added:

- Submit an extract request
- Download an extract
- Find out information about an extract
- Convenience wrapper for creating an API object (done out of crankiness)

It encompasses documentation, tests, and these features.

### CODE REVIEW

**Comment by TheCedarPrince** [comment]
Hey @00krishna , the good news is is that this branch is ready for review! There are a lot of changes I introduced mostly with regards to documentation and the API for extracts, but the new and improved API is this to submit and work with extracts:

```julia
using IPUMS

api = IPUMSAPI("https://api.ipums.org/", Dict("Authorization" => "Your_Key"))

test_extract_definition = "test/testdata/example_extract_request.json"

res = extract_submit(api, "nhgis", test_extract_definition)

metadata, defn, msg = extract_info(api, res.number, "nhgis")

extract_list(api, "nhgis", "2")
```

Let me know what you think and how it works for you!

The big question here is I don't know how best to test these features. Do you have any ideas/suggestions?

File: src/extracts.jl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+Download files associated with a given IPUMS data extract.
+
+### Arguments
+
+- `api::IPUMSAPI` -- An `IPUMSAPI` object to establish connection details.
+
+- `extract_number::Int` -- extract ID assigned to the IPUMS data extract.
+
+- `collection::String` -- What IPUMS collection to be queried for the extract (options could include `"nhgis"`, `"usa"`, etc. corresponding to IPUMS NHGIS or NHGIS USA databases).
```
Comment (00krishna) [inline src/extracts.jl:25]: Should we include an actual list of the databases that we can access. I believe we can access "nhgis", "usa", and "cps" through the api right now.

File: src/piracy.jl [hunk 2]
```diff
@@ -20,3 +19,29 @@ end
 function OpenAPI.from_json(a::Vector{IPUMS.Shapefile}, b)
     [IPUMS.Shapefile(s["name"], s["year"], s["geographicLevel"], s["extent"], s["basis"], s["sequence"]) for s in b["data"]]
 end
+
+"""
+
+This is a pirated method that supports the `extract_list` method in returning additional information about `page_size`, `page_number`, and generated URLs.
```
Comment (00krishna) [inline src/piracy.jl:25]: Should we mention what the resolution to this issue is, in the comment. What does an unpirated version of the method involve? Just to make sure that we know how to fix the problem when we return to it later.

**Comment by 00krishna** [comment]
Hey Jacob. I looked over the code and it look good. I just made a few small comments about content, but nothing major. 

The api looks good. I suppose as we practice with it, we can always polish it. 

In terms of testing, that is a good question. I think the download functionality itself is tested by the `OpenAPI.jl` package folks. You could try and save a local file to the test directory, and then try to download locally? That is one way to do it. Meaning you really only need to test everything after the download.

File: src/extracts.jl (hunk 1 above)
Comment (TheCedarPrince) [reply src/extracts.jl:25]: Oh I do with the `ipums_data_collections` function. Then I was thinking about adding this into the docs.

File: src/piracy.jl (hunk 2 above)
Comment (TheCedarPrince) [reply src/piracy.jl:25]: Good idea. I'll add a comment in the docstring.

**Pull request merged**